File: semana_7/BrunOrtega/quicksort_iterative.py
def quicksort_iterative(arr):
    stack = [(0, len(arr) - 1)]

    while stack:
        low, high = stack.pop()
        if low < high:
            pivot = partition(arr, low, high)
            stack.append((low, pivot - 1))
            stack.append((pivot + 1, high))

    return arr

def partition(arr, low, high):
    pivot = arr[low]
    i = low + 1
    j = high

    while i <= j:
        while i <= high and arr[i] <= pivot:
            i += 1
        while j >= low and arr[j] > pivot:
            j -= 1

        if i <= j:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1

    arr[low], arr[j] = arr[j], arr[low]
    return j

# Example usage
arr = [5, 2, 8, 3, 1, 6, 4]
arr = quicksort_iterative(arr)
print("Sorted array:", arr)

File: semana_7/ivan_peralta/QuickSortIterative.java
import java.util.Stack;

public class QuickSortIterative {

    public static void main(String[] args) {
        int[] arr = { 10, 7, 8, 9, 1, 5 };
        int n = arr.length;

        quickSort(arr, 0, n - 1);
        printArray(arr);
    }

    static void quickSort(int[] arr, int low, int high) {
        Stack<int[]> stack = new Stack<>();
        stack.push(new int[] { low, high });

        while (!stack.isEmpty()) {
            int[] range = stack.pop();
            low = range[0];
            high = range[1];

            int pi = partition(arr, low, high);

            if (pi - 1 > low) {
                stack.push(new int[] { low, pi - 1 });
            }

            if (pi + 1 < high) {
                stack.push(new int[] { pi + 1, high });
            }
        }
    }

    static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = (low - 1);
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    static void printArray(int[] arr) {
        for (int i : arr) {
            System.out.print(i + " ");
        }
        System.out.println();
    }
}
